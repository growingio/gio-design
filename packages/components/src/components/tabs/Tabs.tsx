/* eslint-disable no-underscore-dangle */
import React, { useMemo, useState } from 'react';
import classNames from 'classnames';
import toArray from 'rc-util/lib/Children/toArray';
import { isNil } from 'lodash';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import TabNav from '../tab-nav';
import TabPane from './TabPane';
import { TabProps, TabPaneProps } from './interface';

const Tabs = (props: TabProps, ref: React.Ref<HTMLDivElement>) => {
  const {
    type = 'block',
    size = 'large',
    children,
    prefixCls: customizePrefixCls,
    className,
    activeKey,
    defaultActiveKey = '',
    style,
    onTabClick,
    onChange,
  } = props;
  const [localActiveKey, setLocalActiveKey] = useState<string | number>(activeKey || defaultActiveKey);
  const prefixCls = usePrefixCls('tabs', customizePrefixCls);
  const classString = classNames(prefixCls, className, {
    [`${prefixCls}-${type}`]: true,
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-md`]: size === 'middle',
    [`${prefixCls}-lg`]: size === 'large',
  });

  const [tabNav, tabPane] = useMemo(() => {
    const _tabItem: JSX.Element[] = [];
    const _tabPane = toArray(children).map((node: React.ReactElement<TabPaneProps>) => {
      if (React.isValidElement(node) && node.type === TabPane) {
        const { tab, className: paneClassName, disabled, style: paneStyle, ...restProps } = node.props;
        _tabItem.push(
          <TabNav.Item key={node.key} disabled={disabled}>
            {tab}
          </TabNav.Item>
        );
        return (
          <TabPane
            prefixCls={prefixCls}
            className={classNames(paneClassName, {
              [`${prefixCls}-tabpane-active`]: localActiveKey === node.key,
            })}
            key={node.key as string | number | undefined}
            style={localActiveKey === node.key ? paneStyle : { ...paneStyle, display: 'none' }}
            {...restProps}
          />
        );
      }
      return null;
    });
    return [_tabItem, _tabPane];
  }, [children, localActiveKey, prefixCls]);

  const tabNavKeys = useMemo(() => tabNav.map((item) => item.key!), [tabNav]);
  useMemo(() => {
    if (!tabNavKeys.includes(localActiveKey)) {
      setLocalActiveKey(tabNavKeys[0]);
    }
  }, [localActiveKey, tabNavKeys]);

  useMemo(() => {
    if (!isNil(activeKey) && tabNavKeys.includes(activeKey)) {
      setLocalActiveKey(activeKey);
    }
  }, [activeKey, tabNavKeys]);

  return (
    <div className={classString} ref={ref} style={style}>
      <TabNav
        size={size}
        type={type}
        activeKey={localActiveKey}
        onTabClick={onTabClick}
        onChange={(_key: string | number) => {
          if (isNil(activeKey)) {
            setLocalActiveKey(_key);
          }
          onChange?.(_key);
        }}
      >
        {tabNav}
      </TabNav>
      <div className={`${prefixCls}-content`}>{tabPane}</div>
    </div>
  );
};

export default React.forwardRef(Tabs);
