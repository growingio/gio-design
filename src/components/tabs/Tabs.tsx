import React, { useEffect, useMemo } from 'react';
import classNames from 'classnames';
import toArray from 'rc-util/lib/Children/toArray';
import { isNil } from 'lodash';
import usePrefixCls from '../../utils/hooks/use-prefix-cls';
import TabNav from '../tab-nav';
import TabPane from './TabPane';
import { TabProps, TabPaneProps } from './interface';
import useControlledState from '../../utils/hooks/useControlledState';

const Tabs = (props: TabProps, ref: React.Ref<HTMLDivElement>) => {
  const {
    type = 'block',
    size = 'large',
    children,
    prefixCls: customizePrefixCls,
    className,
    activeKey,
    defaultActiveKey = '',
    style,
    onTabClick,
    onChange,
  } = props;
  const [localActiveKey, setLocalActiveKey] = useControlledState<string>(activeKey, defaultActiveKey);
  const prefixCls = usePrefixCls('tabs', customizePrefixCls);
  const classString = classNames(prefixCls, className, {
    [`${prefixCls}-${type}`]: true,
    [`${prefixCls}-sm`]: size === 'small',
    [`${prefixCls}-md`]: size === 'middle',
    [`${prefixCls}-lg`]: size === 'large',
  });

  const [tabNavKeys, tabNav, tabPane] = useMemo(() => {
    const _tabNavKeys: string[] = [];
    const _tabItem: JSX.Element[] = [];
    const _tabPane =
      toArray(children)
      .filter(node => React.isValidElement(node) && node.type === TabPane)
      .map((node: React.ReactElement<TabPaneProps>, index) => {
        const { tab, className: paneClassName, disabled, style: paneStyle, ...restProps } = node.props;
        const _key = isNil(node.key) ? index.toString() : node.key.toString();
        _tabNavKeys.push(_key);
        _tabItem.push(
          <TabNav.Item key={_key} disabled={disabled}>
            {tab}
          </TabNav.Item>
        );
        return React.cloneElement(node, {
          prefixCls,
          className: classNames(paneClassName, {
            [`${prefixCls}-tabpane-active`]: localActiveKey === _key,
          }),
          style: localActiveKey === _key ? paneStyle : { ...paneStyle, display: 'none' },
          ...restProps,
        });
      });
    return [_tabNavKeys, _tabItem, _tabPane];
  }, [children, localActiveKey, prefixCls]);

  useEffect(() => {
    if (!tabNavKeys.includes(localActiveKey)) {
      setLocalActiveKey(tabNavKeys[0]);
    }
  }, [localActiveKey, tabNavKeys, setLocalActiveKey]);

  return (
    <div className={classString} ref={ref} style={style}>
      <TabNav
        size={size}
        type={type}
        activeKey={localActiveKey}
        onTabClick={onTabClick}
        onChange={(_key: string) => {
          setLocalActiveKey(_key);
          onChange?.(_key);
        }}
      >
        {tabNav}
      </TabNav>
      <div className={`${prefixCls}-content`}>{tabPane}</div>
    </div>
  );
};

export default React.forwardRef(Tabs);
