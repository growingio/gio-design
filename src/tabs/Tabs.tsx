import React, { DOMAttributes } from 'react';
import classnames from 'classnames';
import { isNil } from 'lodash';
import { usePrefixCls } from '@gio-design/utils';
import { TabProps, TabsProps } from './interface';
import useControlledState from '../utils/hooks/useControlledState';
import Tab from './Tab';
import TabButton from './TabButton';
import { WithCommonProps } from '../utils/interfaces';
import WithRef from '../utils/withRef';
import TabsContext from './context';

export const Tabs = WithRef<
  HTMLDivElement,
  WithCommonProps<TabsProps> & Omit<DOMAttributes<HTMLDivElement>, 'onChange'>
>(
  (
    {
      defaultValue = 0,
      value,
      onChange,
      className: classname,
      children,
      size = 'normal',
      tabListStyle,
      ...restProps
    }: WithCommonProps<TabsProps> & Omit<DOMAttributes<HTMLDivElement>, 'onChange'>,
    ref?
  ) => {
    const [activeValue, setActiveValue] = useControlledState<React.Key>(value, defaultValue);
    const prefixCls = usePrefixCls('tabs');
    const tabClasses = classnames(classname, prefixCls);

    const elementList = React.Children.toArray(children).filter(
      (node) => React.isValidElement(node) && node.type === Tab
    );

    const onClick = (v: React.Key) => {
      setActiveValue(v);
      onChange?.(v);
    };

    const tabs = elementList.map((tab: React.ReactElement<WithCommonProps<TabProps>>, index) => (
      <TabButton
        key={tab.props.value}
        value={tab.props.value || index}
        size={size}
        onClick={onClick}
        prefix={tab.props.prefix}
        active={activeValue === tab.props.value}
        disabled={tab.props.disabled}
      >
        {tab.props.label}
      </TabButton>
    ));

    const tabPanels = elementList.map((tab: React.ReactElement<WithCommonProps<TabProps>>, index) => {
      if (isNil(tab.props.value)) {
        return React.cloneElement(<Tab />, { ...tab.props, value: index });
      }
      return React.cloneElement(<Tab />, tab.props);
    });

    return (
      <TabsContext.Provider value={{ activeValue }}>
        <div className={tabClasses} data-testid="tabs" ref={ref} {...restProps}>
          <div data-testid="tablist" className={`${prefixCls}-tablist`} style={tabListStyle}>
            {tabs}
          </div>
          <div data-testid="tabpanels" className={`${prefixCls}-tabpanels`}>
            {tabPanels}
          </div>
        </div>
      </TabsContext.Provider>
    );
  }
);
Tabs.defaultProps = {
  defaultValue: 0,
  size: 'normal',
};

Tabs.displayName = 'Tabs';

export default Tabs;
