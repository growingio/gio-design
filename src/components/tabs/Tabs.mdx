import { Subtitle, ArgsTable } from '@storybook/addon-docs/blocks';
import Tabs from './Tabs';

# Tabs 标签页

<Subtitle>提供平级的区域将大块内容进行收纳和展现，保持界面整洁、层级清晰。</Subtitle>

## 参数说明
<ArgsTable of={Tabs} />

## 代码演示

### 块状标签导航

一共四种尺寸：L（40px）、M（36px）、S（30px）、XS（24px）。

[Example](/?path=/story/basic-components-avatar--block)

### 线状标签导航
提供平级的区域将大块内容进行收纳和展现，保持界面整洁、层级清晰。

一共三种尺寸：L（40px）、M（36px）、S（30px）。

[Example](/?path=/story/basic-components-avatar--line)
