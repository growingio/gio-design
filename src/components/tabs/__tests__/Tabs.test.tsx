import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Tabs, { TabPane } from '../index';

describe('Testing Tabs', () => {
  const getTabs = (args?: unknown) => (
    <Tabs ref={React.createRef()} {...args}>
      <TabPane tab="我的" key="0">
        111
      </TabPane>
      <TabPane tab="全部" key="1">
        222
      </TabPane>
      <TabPane tab="共享" key="2">
        333
      </TabPane>
      <TabPane disabled tab="预置" key="3">
        444
      </TabPane>
    </Tabs>
  );

  it('should be stable', () => {
    const { asFragment } = render(getTabs());
    expect(asFragment()).toMatchSnapshot();
  });

  it('will be changed after click', () => {
    const { getByText, queryAllByTestId, container } = render(getTabs());
    expect(queryAllByTestId('tabnav-item').length).toBe(4);
    fireEvent.click(getByText('全部'));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('全部');
    expect(container.getElementsByClassName('gio-tabs-tabpane-active')[0].textContent).toBe('222');
  })

  it('should be mount, setProps, unmount with no error', () => {
    expect(() => {
      const { rerender, unmount } = render(getTabs());
      rerender(getTabs({ type: 'line' }));
      rerender(getTabs({ size: 'small' }));
      unmount()
    }).not.toThrow();
  });

  test('props type', () => {
    const { container, rerender } = render(getTabs());
    expect(container.getElementsByClassName('gio-tabs-block')).not.toBe(null);
    rerender(getTabs({ type: 'line' }));
    expect(container.getElementsByClassName('gio-tabs-line')).not.toBe(null);
  });

  test('props size', () => {
    const { container, rerender } = render(getTabs());
    expect(container.getElementsByClassName('gio-tabs-lg')).not.toBe(null);
    rerender(getTabs({ size: 'middle' }));
    expect(container.getElementsByClassName('gio-tabs-md')).not.toBe(null);
    rerender(getTabs({ size: 'small' }));
    expect(container.getElementsByClassName('gio-tabs-sm')).not.toBe(null);
  });

  test('prop defaultActiveKey', () => {
    const { container } = render(getTabs({ defaultActiveKey: '1' }))
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('全部');
  });

  test('prop activeKey', () => {
    const { rerender, container, getByText } = render(getTabs({ activeKey: '1' }));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('全部');
    rerender(getTabs({ activeKey: '2' }));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('共享');
    fireEvent.click(getByText('我的'));
    expect(container.getElementsByClassName('gio-tabnav-item-active')[0].textContent).toBe('共享');
  });

  test('prop onChange and onTabClick', () => {
    const onChange = jest.fn();
    const onTabClick = jest.fn();
    const { getByText } = render(getTabs({ onChange, onTabClick }));
    fireEvent.click(getByText('我的'));
    // not change
    expect(onChange).not.toHaveBeenCalled();
    expect(onTabClick).toHaveBeenCalled();
    fireEvent.click(getByText('全部'));
    expect(onChange).toHaveBeenCalled();
  });

  it('should be render content rightly', () => {
    const { container, getByText } = render(getTabs());
    expect(container.getElementsByClassName('gio-tabs-tabpane-active')[0].textContent).toBe('111');
    fireEvent.click(getByText('全部'));
    expect(container.getElementsByClassName('gio-tabs-tabpane-active')[0].textContent).toBe('222');
  });

  it('only render TabPane', () => {
    const { queryByText } = render(
      <Tabs defaultActiveKey="1">
        <TabPane tab="我的" key="0">
          111
        </TabPane>
        <TabPane tab="全部" key="1">
          222
        </TabPane>
        <TabPane tab="共享" key="2">
          333
        </TabPane>
        <TabPane disabled tab="预置" key="3">
          444
        </TabPane>
        <div>我不会被渲染</div>
      </Tabs>
    );
    expect(queryByText('我不会被渲染')).toBe(null);
  });
});
