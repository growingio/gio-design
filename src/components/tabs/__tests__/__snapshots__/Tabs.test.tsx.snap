// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Tabs should be stable 1`] = `
<DocumentFragment>
  <div
    class="gio-tabs gio-tabs-block gio-tabs-lg"
  >
    <div
      class="gio-tabnav gio-tabnav-block gio-tabnav-lg"
    >
      <div
        class="gio-tabnav-item gio-tabnav-item-active"
        data-testid="tabnav-item"
      >
        <div
          class="gio-tabnav-item-btn"
        >
          我的
        </div>
      </div>
      <div
        class="gio-tabnav-item"
        data-testid="tabnav-item"
      >
        <div
          class="gio-tabnav-item-btn"
        >
          全部
        </div>
      </div>
      <div
        class="gio-tabnav-item"
        data-testid="tabnav-item"
      >
        <div
          class="gio-tabnav-item-btn"
        >
          共享
        </div>
      </div>
      <div
        class="gio-tabnav-item gio-tabnav-item-disabled"
        data-testid="tabnav-item"
        disabled=""
      >
        <div
          class="gio-tabnav-item-btn"
        >
          预置
        </div>
      </div>
      <div
        class="gio-tabnav-ink-bar"
        style="left: 0px; width: 0px;"
      />
    </div>
    <div
      class="gio-tabs-content"
    >
      <div
        class="gio-tabs-tabpane gio-tabs-tabpane-active"
        style=""
      >
        111
      </div>
      <div
        class="gio-tabs-tabpane"
        style="display: none;"
      >
        222
      </div>
      <div
        class="gio-tabs-tabpane"
        style="display: none;"
      >
        333
      </div>
      <div
        class="gio-tabs-tabpane"
        style="display: none;"
      >
        444
      </div>
    </div>
  </div>
</DocumentFragment>
`;
